libs/pika/debugging/include/pika/debugging/demangle_helper.hpp

```diff
@@ -8,103 +8,76 @@
+#endif
+
 // --------------------------------------------------------------------
 namespace pika::debug::detail {
-    template <typename T>
+    // default : use built-in typeid to get the best info we can
+    template <typename T, typename Enabled = std::false_type>
     struct demangle_helper
     {
         char const* type_id() const { return typeid(T).name(); }
     };
-}    // namespace pika::debug::detail
-
-#if defined(__GNUG__)
 
-# include <cxxabi.h>
-# include <memory>
-# include <stdlib.h>
-
-// --------------------------------------------------------------------
-// if available : demangle an arbitrary c++ type using gnu utility
-// --------------------------------------------------------------------
-namespace pika::debug::detail {
+    // if available : demangle an arbitrary c++ type using gnu utility
     template <typename T>
-    class cxxabi_demangle_helper
+    struct demangle_helper<T, std::true_type>
     {
-    public:
-        cxxabi_demangle_helper()
+        demangle_helper()
           : demangled_{abi::__cxa_demangle(typeid(T).name(), nullptr, nullptr, nullptr), std::free}
         {
         }
 
         char const* type_id() const { return demangled_ ? demangled_.get() : typeid(T).name(); }
 
     private:
-        std::unique_ptr<char, void (*)(void*)> demangled_;
+        std::unique_ptr<char, decltype(&std::free)> demangled_;
     };
-}    // namespace pika::debug::detail
-
-#else
 
-namespace pika::debug::detail {
     template <typename T>
-    using cxxabi_demangle_helper = demangle_helper<T>;
+    using cxx_type_id = demangle_helper<T, cxxabi_supported__>;
 }    // namespace pika::debug::detail
 
-#endif
-
-///////////////////////////////////////////////////////////////////////////////
-namespace pika::debug::detail {
-    template <typename T>
-    struct type_id
-    {
-        static demangle_helper<T> typeid_;
-    };
-
-    template <typename T>
-    demangle_helper<T> type_id<T>::typeid_ = demangle_helper<T>();
-
-#if defined(__GNUG__)
-    template <typename T>
-    struct cxx_type_id
-    {
-        static cxxabi_demangle_helper<T> typeid_;
-    };
-
-    template <typename T>
-    cxxabi_demangle_helper<T> cxx_type_id<T>::typeid_ = cxxabi_demangle_helper<T>();
-#else
-    template <typename T>
-    using cxx_type_id = type_id<T>;
-#endif
-
-    // --------------------------------------------------------------------
-    // print type information
-    // usage : std::cout << print_type<args...>("separator")
-    // separator is appended if the number of types > 1
-    // --------------------------------------------------------------------
-    template <typename T = void>
+// --------------------------------------------------------------------
+// print type information
+// usage : std::cout << debug::print_type<args...>("separator")
+// separator is appended if the number of types > 1
+// --------------------------------------------------------------------
+namespace pika::debug {
```

Comment:
Is this meant to be user-facing? If not, we should keep it in `pika::debug::detail` (we can shorten it to `pika::detail` as well).

Reply:
yes, I wanted `pika::debug::print_xxx` to be in the pika::debug namespace as these are the user facing apis that people might use.
Just pushing a new version that I think resolves the abi namespace issue cleanly. Don't know why I didn't do it first time around, but that's why we have these annoying reviews ....